src/contracts/engine/transactions/sellAndBurnMln.test.ts

```diff
@@ -0,0 +1,217 @@
+import { BigNumber } from 'bignumber.js';
+import { appendDecimals } from '@melonproject/token-math/token';
+import { createQuantity } from '@melonproject/token-math/quantity';
+import { getPrice } from '@melonproject/token-math/price';
+import { initTestEnvironment, getGlobalEnvironment } from '~/utils/environment';
+import { Address } from '~/utils/types';
+import { deploy as deployEngine, sellAndBurnMln } from '..';
+import {
+  deploy as deployToken,
+  approve,
+  getToken,
+  balanceOf,
+} from '~/contracts/dependencies/token';
+import { deploy as deployFeed, update } from '~/contracts/prices';
+import {
+  increaseTime,
+  deployAndGetContract,
+  getContract,
+} from '~/utils/solidity';
+import { Contracts } from '~/Contracts';
+
+const shared: any = {};
+
+// TODO: use the actual sellAndBurnMln typescript function rather than testing like this
+beforeAll(async () => {
+  await initTestEnvironment();
+  shared.env = getGlobalEnvironment();
+  shared.accounts = await shared.env.eth.getAccounts();
+  const wethAddress = await deployToken('ETH');
+  shared.mln = await deployAndGetContract('dependencies/token/BurnableToken', [
+    'MLN',
+    18,
+    '',
+  ]);
+  shared.weth = await getContract(Contracts.StandardToken, wethAddress);
+  shared.version = await deployAndGetContract('version/MockVersion');
+  const feedAddress = await deployFeed(await getToken(wethAddress));
+  shared.feed = await getContract(Contracts.TestingPriceFeed, feedAddress);
+  shared.delay = 30 * 24 * 60 * 60;
+  shared.engineAddress = await deployEngine(
+    shared.version.options.address,
+    feedAddress,
+    shared.delay,
+    shared.mln.options.address,
+  );
+  shared.priceSource = await getContract(
+    Contracts.TestingPriceFeed,
+    feedAddress,
+  );
+  shared.engine = getContract(Contracts.Engine, shared.engineAddress);
+  await shared.feed.methods
+    .update(
+      [shared.weth.options.address, shared.mln.options.address],
+      [
+        new BigNumber(10 ** 18).toFixed(),
+        new BigNumber(0.0588 * 10 ** 18).toFixed(),
+      ],
+    )
+    .send({ from: shared.env.wallet.address, gas: 8000000 });
+});
+
+test('directly sending eth fails', async () => {
+  await expect(
+    shared.env.eth.sendTransaction({
+      from: shared.env.wallet.address,
+      to: shared.engine.options.address,
+      value: 1000000,
+    }),
+  ).rejects.toThrow('revert');
+});
+
+test('eth sent via contract selfdestruct is not tracked', async () => {
+  const sendEth = 10 ** 18;
```

Comment:
Be carefull with these numbers. `10 ** 18` is already greater than `Number.MAX_SAFE_INTEGER (9007199254740991)`. Better to use `token-math/big-integer`.

Reply:
Fair; goes along with using BigInteger above instead of this